File: packages/web/packages/avatar-engine/src/math/mat.ts
/**
 * A 2x3 affine, laid out the way SVG and CoreGraphics both spell it:
 *
 *   | a c e |
 *   | b d f |
 *   | 0 0 1 |
 *
 * The engine composes every node's transform down to one of these and hands the
 * renderer the result, so no renderer ever has to know what a pivot is.
 */
export type Mat = readonly [number, number, number, number, number, number];

export const IDENTITY: Mat = [1, 0, 0, 1, 0, 0];

const DEG = Math.PI / 180;

export function multiply(m: Mat, n: Mat): Mat {
  const [a1, b1, c1, d1, e1, f1] = m;
  const [a2, b2, c2, d2, e2, f2] = n;
  return [
    a1 * a2 + c1 * b2,
    b1 * a2 + d1 * b2,
    a1 * c2 + c1 * d2,
    b1 * c2 + d1 * d2,
    a1 * e2 + c1 * f2 + e1,
    b1 * e2 + d1 * f2 + f1,
  ];
}

export function applyPoint(m: Mat, x: number, y: number): [number, number] {
  return [m[0] * x + m[2] * y + m[4], m[1] * x + m[3] * y + m[5]];
}

export function fromTransform(t: {
  x?: number;
  y?: number;
  rotation?: number;
  scaleX?: number;
  scaleY?: number;
  pivot?: readonly [number, number];
}): Mat {
  const sx = t.scaleX ?? 1;
  const sy = t.scaleY ?? 1;
  const r = (t.rotation ?? 0) * DEG;
  const cos = Math.cos(r);
  const sin = Math.sin(r);
  const [px, py] = t.pivot ?? [0, 0];

  // rotate+scale about the pivot: T(p) . R . S . T(-p)
  const a = cos * sx;
  const b = sin * sx;
  const c = -sin * sy;
  const d = cos * sy;
  const e = px - (a * px + c * py) + (t.x ?? 0);
  const f = py - (b * px + d * py) + (t.y ?? 0);
  return [a, b, c, d, e, f];
}

File: packages/apple/AvatarAnimationEngine/Sources/Math/Mat.swift
import Foundation

/// A 2x3 affine, laid out the way SVG and CoreGraphics both spell it:
///
///     | a c e |
///     | b d f |
///     | 0 0 1 |
///
/// The engine composes every node's transform down to one of these and hands
/// the renderer the result, so no renderer ever has to know what a pivot is.
///
/// This is a plain struct of `Double` rather than a `CGAffineTransform` for one
/// reason: `CGAffineTransform`'s storage is `CGFloat`, which is `Float` on some
/// architectures and would quietly cost us the 1e-6 golden tolerance. The
/// bridge below is the only place the two meet.
public struct Mat: Equatable, Sendable {
    public var a: Double
    public var b: Double
    public var c: Double
    public var d: Double
    public var e: Double
    public var f: Double

    public init(a: Double, b: Double, c: Double, d: Double, e: Double, f: Double) {
        self.a = a; self.b = b; self.c = c; self.d = d; self.e = e; self.f = f
    }

    public static let identity = Mat(a: 1, b: 0, c: 0, d: 1, e: 0, f: 0)

    /// Parent on the left, child on the right — the same argument order as the
    /// web's `multiply(parent, child)`.
    public static func * (m: Mat, n: Mat) -> Mat {
        Mat(
            a: m.a * n.a + m.c * n.b,
            b: m.b * n.a + m.d * n.b,
            c: m.a * n.c + m.c * n.d,
            d: m.b * n.c + m.d * n.d,
            e: m.a * n.e + m.c * n.f + m.e,
            f: m.b * n.e + m.d * n.f + m.f
        )
    }

    public func apply(x: Double, y: Double) -> (x: Double, y: Double) {
        (a * x + c * y + e, b * x + d * y + f)
    }

    /// Rotation is in degrees and the pivot is in absolute design units, both
    /// matching the config format. Composition is
    /// `T(x,y) . T(p) . R . S . T(-p)`.
    public static func from(
        x: Double = 0,
        y: Double = 0,
        rotation: Double = 0,
        scaleX: Double = 1,
        scaleY: Double = 1,
        pivot: (Double, Double) = (0, 0)
    ) -> Mat {
        let r = rotation * .pi / 180
        let cos = Foundation.cos(r)
        let sin = Foundation.sin(r)
        let (px, py) = pivot
        let a = cos * scaleX
        let b = sin * scaleX
        let c = -sin * scaleY
        let d = cos * scaleY
        return Mat(
            a: a, b: b, c: c, d: d,
            e: px - (a * px + c * py) + x,
            f: py - (b * px + d * py) + y
        )
    }
}
